Issue #3045492 by mtodor: Backend performance testing

mtodor opened the pull request:
Make sure these boxes are checked before submitting your pull request - thank you!

- [ ] All coding styles are fulfilled. ([How to check for cs issues?](https://github.com/BurdaMagazinOrg/thunder-dev-tools/blob/master/README.md#code-style-guidelines))
- [ ] All tests are running locally. ([How to run the test?](https://github.com/BurdaMagazinOrg/thunder-distribution/blob/develop/docs/development.md#how-to-run-the-tests))
- [ ] Necessary update hooks are provided.
- [ ] User roles have correct access for new introduced permission.
- [ ] Every thunder module has a README.md in its root. Follow [this guidelines](https://www.drupal.org/node/2181737), but we don't need every topic.
- [ ] Code is covered with well-balanced amount of inline comments.

If you are really awesome, then your feature is covered by additional tests. Well done!

mtodor committed:
Initial inegration for Elastic APM for performance testing

mtodor committed:
Improved base test, additional commands and added required fixtures

mtodor committed:
Added window resize

mtodor committed:
Added support for adding tags for transaction

mtodor committed:
Fixed coding styles

mtodor committed:
Added support for configurable Elastic APM endpoint and domain

mtodor committed:
Added branch tag for better data filtering

mtodor committed:
Refactor paragraphs creation

mtodor committed:
Moved Elastic APM inclusion from globals to test file

mtodor committed:
Test naming unification

mtodor committed:
Fix issue with importing of Elasti APM module

mtodor committed:
Added image upload and big code styling adjustments

mtodor committed:
Fixed images with size over 2MB

mtodor committed:
Removed images from fixtures and removed image uploads from test.

mtodor committed:
Merge branch '8.x-2.x' into feature/3045492-backend-performance-testing

mtodor committed:
Added randomization for SEO title

mtodor committed:
Removed path require

mtodor commented:
We will move this functionality into a separate module.

mtodor committed:
Moved Thunder RUM in separate module

mtodor committed:
Added thunder performance measurement module

mtodor committed:
Merge branch '8.x-2.x' into feature/3045492-backend-performance-testing

mtodor committed:
Removed confusing command: performanceMark

mtodor commented:
I have moved only required libraries for Browser Performance Measurement to separate module.

mtodor committed:
Added AWS buid spec

mtodor committed:
Merge branch '8.x-2.x' into feature/3045492-backend-performance-testing

mtodor committed:
Rename AWS buid spec

mtodor committed:
Removed AWS build integration

mtodor committed:
Merge branch '8.x-2.x' into feature/3045492-backend-performance-testing

dbosen committed:
Merge branch '8.x-2.x' into feature/3045492-backend-performance-testing

alexpott commented on tests/src/Nightwatch/Commands/paragraphsAddEmpty.js:
```diff
@@ -0,0 +1,43 @@
+ *   The 'browser' object.
+ */
+
+exports.command = function paragraphsAddEmpty(fieldName, type, position) {
```
This is only used in the paragraphsAdd command - is there a way to not make it part of the global commands but just something in scope there.

mtodor replied:
After split of `paragraphsAdd.js` into multiple commands, this command is used in several places.

alexpott commented on tests/src/Nightwatch/Commands/paragraphsAdd.js:
```diff
@@ -0,0 +1,155 @@
+  const newParagraphPosition = position * 2;
+
+  // Text Paragraph.
+  if (type === "text") {
```
I think that rather than this method being a big switch on type that having addTextParagraph, addImageParagraphs, addInstagramParagraph is neater and more expressive.

alexpott replied:
Also why paragraphsAdd() and not addParagraph()? I guess maybe to group commands. Maybe we could do something like paragraphs.addText()... and performance.startMark() - I dunno my JS-fu is not good enough to know off-hand.

mtodor replied:
I didn't find in documentation of Nightwatch if I can group commands in folders, but after checking code, I found it's possible. So, I have restructured all commands in sub-folders.

alexpott commented on tests/src/Nightwatch/Commands/performanceMarkEnd.js:
```diff
@@ -0,0 +1,42 @@
+
+    browser.setCookie({
+      domain: browser.apmDomain,
+      expiry: 3533274000,
```
This is optional - do we really need it? Also why Thursday, 18 December 2081 09:00:00

mtodor replied:
Yes, we do not need this option. I'll remove it.

And why `3533274000`, well that would be my 100th birthday. 😄

alexpott commented on tests/src/Nightwatch/Commands/performanceMarkStart.js:
```diff
@@ -0,0 +1,34 @@
+
+    browser.setCookie({
+      domain: browser.apmDomain,
+      expiry: 3533274000,
```
As above - and the rest below - is this necessary?

mtodor replied:
👍

alexpott commented on tests/src/Nightwatch/Commands/thunderFillCKEditor.js:
```diff
@@ -0,0 +1,33 @@
+ */
+
+/* eslint-disable func-names */
+exports.command = function thunderFillCKEditor(selector, value) {
```
Not sure about the thunder prefix here

mtodor replied:
I'll remove it. I have prefixed it because I didn't want to have potential command collisions with core commands if they are introduced one day. But, if that happens, we will fix it.

alexpott commented on tests/src/Nightwatch/Commands/thunderScrollIntoView.js:
```diff
@@ -0,0 +1,49 @@
+ */
+
+/* eslint-disable func-names */
+exports.command = function thunderScrollIntoView(selector) {
```
I think we can use https://nightwatchjs.org/api/moveToElement.html here instead - no?

mtodor replied:
Problem with existing `move to element` functionalities is that they move an element into view of the browser window, but overlay elements are still there. So when the driver does click it clicks on overlay element and the test fails. That's why we are using scroll into view - what is actually doing scroll into the middle of the view.

alexpott commented on tests/src/Nightwatch/Commands/thunderSelectTag.js:
```diff
@@ -0,0 +1,27 @@
+/**
+ * @file
+ * Sets tag for element. It can be existing tag or new one.
```
I think this could be more explicit about what we're dealing with. I think this is about taxonomy terms but select and tag are very very overloaded language.

mtodor replied:
It's widget specific, so I'll move it into widget related namespace.

mtodor commented:
@alexpott I have addressed all your comments.

I have moved PR to the new repository. You can find it here: https://github.com/thunder/thunder-distribution/pull/19

First commit is code that you have reviewed. Commits after are related to your review and some Elastic APM deprecation warnings.

I'll close this PR since it's obsolete.
